homeassistant/components/tedee/binary_sensor.py

```diff
@@ -0,0 +1,78 @@
+"""Tedee sensor entities."""
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from pytedee_async import TedeeLock
+from pytedee_async.lock import TedeeLockState
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorDeviceClass,
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EntityCategory
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .entity import TedeeDescriptionEntity
+
+
+@dataclass(frozen=True, kw_only=True)
+class TedeeBinarySensorEntityDescription(
+    BinarySensorEntityDescription,
+):
+    """Describes Tedee binary sensor entity."""
+
+    is_on_fn: Callable[[TedeeLock], bool | None]
+
+
+ENTITIES: tuple[TedeeBinarySensorEntityDescription, ...] = (
+    TedeeBinarySensorEntityDescription(
+        key="charging",
+        device_class=BinarySensorDeviceClass.BATTERY_CHARGING,
+        is_on_fn=lambda lock: lock.is_charging,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+    TedeeBinarySensorEntityDescription(
+        key="semi_locked",
+        translation_key="semi_locked",
+        is_on_fn=lambda lock: lock.state == TedeeLockState.HALF_OPEN,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+    TedeeBinarySensorEntityDescription(
+        key="pullspring_enabled",
+        translation_key="pullspring_enabled",
+        is_on_fn=lambda lock: lock.is_enabled_pullspring,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Tedee sensor entity."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+
+    for entity_description in ENTITIES:
+        async_add_entities(
```

Comment:
Please collect all the entities first and then make a single call to `async_add_entities`. Just include the iteration of the entity descriptions in the list comprehension below.

Reply:
Additionally, we can make it a generator expression instead of a list comprehension.

Reply:
addressed in #109858